danmaku/ui/commonMain/DanmakuHostState.kt

```diff
@@ -0,0 +1,393 @@
+            FloatingDanmakuTrack(
+                trackIndex = index,
+                frameTimeNanos = elapsedFrameTimeNanoState,
+                trackHeight = trackHeightState,
+                screenWidth = hostWidthState, 
+                speedPxPerSecond = newFloatingTrackSpeed,
+                safeSeparation = newFloatingTrackSafeSeparation,
+                baseTextLength = floatingBaseTextLengthForSpeed,
+                speedMultiplier = floatingSpeedMultiplierState,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        floatingTrack.forEach {
+            it.speedPxPerSecond = newFloatingTrackSpeed
+            it.safeSeparation = newFloatingTrackSafeSeparation
+            it.baseTextLength = floatingBaseTextLengthForSpeed
+        }
+        topTrack.setTrackCountImpl(if (config.enableTop) count else 0) { index ->
+            FixedDanmakuTrack(
+                trackIndex = index,
+                frameTimeNanos = elapsedFrameTimeNanoState,
+                trackHeight = trackHeightState,
+                screenWidth = hostWidthState,
+                screenHeight = hostHeightState,
+                fromBottom = false,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        bottomTrack.setTrackCountImpl(if (config.enableBottom) count else 0) { index ->
+            FixedDanmakuTrack(
+                trackIndex = index,
+                frameTimeNanos = elapsedFrameTimeNanoState,
+                trackHeight = trackHeightState,
+                screenWidth = hostWidthState,
+                screenHeight = hostHeightState,
+                fromBottom = true,
+                onRemoveDanmaku = { removed -> mutablePresentDanmaku.remove(removed) }
+            )
+        }
+        invalidate()
+    }
+
+    private fun <T : DanmakuTrack> MutableList<T>.setTrackCountImpl(
+        count: Int,
+        newInstance: (index: Int) -> T,
+    ) {
+        when {
+            size == count -> return
+            // 清除 track 的同时要把 track 里的 danmaku 也要清除
+            count < size -> repeat(size - count) { removeLast().clearAll() }
+            else -> addAll(List(count - size) { newInstance(size + it) })
+        }
+    }
+
+    @UiThread
+    internal suspend fun interpolateFrameLoop() {
+        coroutineScope {
+            launch {
+                var currentFrameTimeNanos = withFrameNanos {
+                    // 使用了这一帧来获取时间, 需要补偿平均帧时间
+                    elapsedFrameTimeNanos += avgFrameTimeNanos.avg()
+                    it
+                }
+
+                while (true) {
+                    withFrameNanos { nanos ->
+                        val delta = nanos - currentFrameTimeNanos
+                        elapsedFrameTimeNanos += delta
+                        avgFrameTimeNanos += delta
+                        currentFrameTimeNanos = nanos
+                    }
```

Comment:
目前我没有看到足够的理由来维护 `elapsedFrameTimeNanos` 和 avgFrameTimeNanos

use `nanos` strictly:

```suggestion
                    withFrameNanos { nanos ->
                        currentFrameTimeNanos = nanos
                    }
```

Reply:
在暂停和恢复的时候有用，如果直接用 withFrame 的时间将导致弹幕跳动，需要花费额外的精力重新计算位置